src/typesetting.rs

```diff
@@ -0,0 +1,205 @@
+
+    /// HTML definitions are used to replace a token with a piece of HTML syntax.
+    /// This version will generally be used for the GIF rendering version of the web pages.
+    /// Each entry has the form `(token, replacement)`.
+    /// ```text
+    /// htmldef "ph" as "<IMG SRC='_varphi.gif' WIDTH=11 HEIGHT=19 ALT=' ph' TITLE='ph'>";
+    /// ```
+    pub html_defs: HashMap<Token, Token>,
+
+    /// HTML definitions are used to replace a token with a piece of HTML syntax.
+    /// This version will generally be used for the unicode rendering version of the web pages.
+    /// Each entry has the form `(token, replacement)`.
+    /// ```text
+    /// althtmldef "ph" as "<SPAN CLASS=wff STYLE='color:blue'>&#x1D711;</SPAN>";
+    /// ```
+    pub alt_html_defs: HashMap<Token, Token>,
+
+    /// A piece of HTML to give the variable color key. All `htmlvarcolor` directives are given
+    /// separately here, but they are logically concatenated with spaces for rendering.
+    /// ```text
+    /// htmlvarcolor '<SPAN CLASS=wff STYLE="color:blue;font-style:normal">wff</SPAN> '
+    ///   + '<SPAN CLASS=setvar STYLE="color:red;font-style:normal">setvar</SPAN> '
+    ///   + '<SPAN CLASS=class STYLE="color:#C3C;font-style:normal">class</SPAN>';
+    /// ```
+    pub html_var_color: Vec<Token>,
+
+    /// The title of the generated HTML page.
+    /// ```text
+    /// htmltitle "Metamath Proof Explorer";
+    /// ```
+    pub html_title: Option<Token>,
+
+    /// The link to the home page in the generated HTML page.
+    /// ```text
+    /// htmlhome '<A HREF="mmset.html"><FONT SIZE=-2 FACE=sans-serif>' +
+    ///     '<IMG SRC="mm.gif" BORDER=0 ALT='  +
+    ///     '"Home" HEIGHT=32 WIDTH=32 ALIGN=MIDDLE STYLE="margin-bottom:0px">' +
+    ///     'Home</FONT></A>';
+    /// ```
+    pub html_home: Option<Token>,
+
+    /// The relative path from the unicode version to the GIF version. Used for cross references.
+    /// (This is a set.mm specific hack.)
+    /// ```text
+    /// htmldir "../mpegif/";
+    /// ```
+    pub html_dir: Option<Token>,
+
+    /// The relative path from the GIF version to the unicode version. Used for cross references.
+    /// (This is a set.mm specific hack.)
+    /// ```text
+    /// althtmldir "../mpeuni/";
+    /// ```
+    pub alt_html_dir: Option<Token>,
+
+    /// Optional file where bibliographic references are kept.
+    /// ```text
+    /// htmlbibliography "mmset.html";
+    /// ```
+    pub html_bibliography: Option<Token>,
+
+    /// Custom CSS to be placed in the header of generated files.
+    /// Note that any `\n` escapes are not yet replaced by newlines in this `html_css` variable;
+    /// library consumers are responsible for performing this replacement.
+    /// ```text
+    /// htmlcss '<STYLE TYPE="text/css">\n' +
+    ///   '</STYLE>\n' +
+    ///   '<LINK href="mmset.css" title="mmset"\n' +
+    ///   '    rel="stylesheet" type="text/css">\n';
+    /// ```
+    pub html_css: Option<Token>,
```

Comment:
The `html_css` field seems to have actual `\n` strings inside, it shall be replaced by line feeds when parsing it (otherwise browsers don't understand it and assume this is part of HTML `body`, not `head`).

Reply:
I don't think this is terribly blocking, we could still merge and fix later.

Reply:
Yes, that's intended behavior, as documented in the comment.

Reply:
Oh, sorry, I shall have read the comment until the end! :)
Obviously I would rather have assumed that this facility would be provided by the library!